Fix ONNX loading in issues opencv#17516, opencv#17531

sl-sergei opened the pull request:
**Merge with extra:** opencv/opencv_extra#793

resolves #17516
resolves #17531
 
Also resolves untracked problem with constant input for Gemm layers

```
opencv_extra=testdata_for_17516_17531

force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.4.0:16.04
build_image:Custom Win=openvino-2020.2.0
build_image:Custom Mac=openvino-2020.3.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

asmorkalov commented:
@sl-sergei Please add test for the failed case.

dkurt commented:
Please target this patch to 3.4 branch

asmorkalov commented:
@sl-sergei  Friendly reminder about test.

dkurt commented on modules/dnn/src/onnx/onnx_importer.cpp:
```diff
@@ -902,6 +902,19 @@ void ONNXImporter::populateNet(Net dstNet)
+                opencv_onnx::NodeProto proto;
+                proto.add_output(constParams.name);
+                addLayer(dstNet, constParams, proto, layer_id, outShapes);
+            }
```
Please add a test which covers this code. Thanks!

dkurt commented:
Please avoid merge commits. We'd like to recommend to do rebase instead.

seems like there are failed tests: https://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux/builds

dkurt commented:
Please take a look at failed tests: https://pullrequest.opencv.org/buildbot/builders/precommit_custom_windows/builds/1317

dkurt commented on modules/dnn/test/test_onnx_importer.cpp:
```diff
@@ -606,6 +606,16 @@ TEST_P(Test_ONNX_layers, Pad2d_Unfused)
+TEST_P(Test_ONNX_layers, MatmulWithTwoInputs)
+{
+    testONNXModels("matmul_with_two_inputs");
+}
```
Please skip the tests for all the versions lower 2020.4: 
2020.3 - failed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_mac/builds/1053
2020.4 - passed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/4279

sl-sergei replied:
> Please skip the tests for all the versions lower 2020.4:
> 2020.3 - failed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_mac/builds/1053
> 2020.4 - passed https://pullrequest.opencv.org/buildbot/builders/precommit_custom_linux/builds/4279

Thank you, I tried to reproduce it locally

dkurt replied:
It's OK just add skip exception (see `applyTestTag`)

dkurt commented:
```
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] Test_ONNX_layers.LinearWithConstant/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_layers.MatmulWithTwoInputs/1, where GetParam() = OCV/OCL_FP16
```
https://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux/builds/25739

sl-sergei commented:
> ```
> [  FAILED  ] 2 tests, listed below:
> [  FAILED  ] Test_ONNX_layers.LinearWithConstant/1, where GetParam() = OCV/OCL_FP16
> [  FAILED  ] Test_ONNX_layers.MatmulWithTwoInputs/1, where GetParam() = OCV/OCL_FP16
> ```
> 
> https://pullrequest.opencv.org/buildbot/builders/precommit_opencl_linux/builds/25739

I see, it looks like it fails sporadically on Intel GPU's with OpenCL. There were no changes since the last successful run

dkurt reviewed (approved):
:+1: Thank you!
